google/cloud/spanner/internal/background_threads_impl.cc

```diff
@@ -0,0 +1,40 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/spanner/internal/background_threads_impl.h"
+
+namespace google {
+namespace cloud {
+namespace spanner {
+inline namespace SPANNER_CLIENT_NS {
+namespace internal {
+
+AutomaticallyCreatedBackgroundThreads::AutomaticallyCreatedBackgroundThreads()
+    : runner_([](grpc_utils::CompletionQueue cq) { cq.Run(); }, cq_) {}
+
+AutomaticallyCreatedBackgroundThreads::
+    ~AutomaticallyCreatedBackgroundThreads() {
+  Shutdown();
+}
+
+void AutomaticallyCreatedBackgroundThreads::Shutdown() {
+  cq_.Shutdown();
+  if (runner_.joinable()) runner_.join();
```

Comment:
How might `runner_` have become detached if we don't provide any access to it?

Reply:
You can call `Shutdown()` twice, or `Shutdown()` and then the destructor.  In both cases the second call would throw:

https://en.cppreference.com/w/cpp/thread/thread/join

Reply:
Calling `join()` (via `Shutdown()` for example) leaves `runner_` in a non-joinable state, in which case `std::thread::join` throws.

Reply:
OK.  (I wonder whether this class should do something to ensure it only ever tries to shutdown the CompletionQueue and join the thread once, but I'm happy if this has the same effect.)